File: K-means.cpp
#include<iostream>
#include<string>
#include<vector>
#include<ctime>
#include<fstream>
#include<algorithm>
#include<numeric>

#define PI 3.151592653


using namespace std;
using std::vector;
using std::string;
using std::fstream;



//函数random生成a-b之间的随机数
int random(int a,int b)
{
	int dis = b - a;
	return rand() % dis + a;
}


//函数rad转度数为弧度
double rad(double a)
{
	return a * PI / 180;
}



//函数distance计算两个经纬度之间的距离
double distance(double longitude1, double latitude1, double longitude2, double latitude2)
{
	double Lat1 = rad(latitude1);
	double Lat2 = rad(latitude2);
	double a = Lat1 - Lat2;
	double b = rad(longitude1) - rad(longitude2);
	double s = 2 * asin(sqrt(pow(sin(a / 2), 2) + cos(Lat1)*cos(Lat2)*pow(sin(b / 2), 2)));
	s = s * 6378137;
	return s;
}



//主函数
int main()
{
 
	ifstream input;                    //读取文件信息
	ofstream output;                   //输出文件信息
	double longistude;                 //共享单车经度
	double latitude;                   //共享单车维度
	vector<double> ivecii;             //暂存每辆单车的经纬度信息
	vector<vector<double>> ivec;       //储存所有单车的经纬度
	vector<int> position;              //储存随机生成的位置点
	vector<vector<vector<double>>> end(500);   //选择生成500个区域

	vector<double>  result;            //储存每个点到所选所有位置点的距离


	vector<double> longistudei;        //储存某个区域所有的经度
	vector<double> latitudei;          //储存某个区域所有的维度


	double a;                          //
	double b;                          //






	input.open("C:\\Users\\Rooobins\\Desktop\\DataXXYY.txt", fstream::in);
	

	//读取每辆单车的位置信息并储存
	if (!input.is_open())
	{
		cout << "Open file error!" << endl;
	}
	else
	{
		for (int i = 0; i != 30557; ++i)
		{
			input >> longistude >> latitude;
			ivecii.push_back(longistude);
			ivecii.push_back(latitude);
			ivec.push_back(ivecii);
			ivecii.clear();
		}
	}
	input.close();


    //生成500个区域的初始位置点
	for (int i = 0; i != 500; ++i)
	{
		position.push_back(random(0, ivec.size()));
	}

	


	for (auto i = 0; i != ivec.size(); ++i)
	{
		for (auto j = 0; j != position.size(); ++j)
		{
			
             result.push_back(distance(ivec[i][0],ivec[i][1],ivec[position[j]][0],ivec[position[j]][1]));
		}

		auto MIN=min_element(result.begin(),result.end());
		int P=*MIN-result.begin();
		end[P].push_back(ivec[i]);
	}


	for(auto i=0;i!=end.size();++i)
	{
		double a1=0;
		double b1=0;
		for(auto j=0;j!=end[i].size();++j)
		{
			a1+=end[i][j][0];
			b1++end[i][j][1];
		}
	}

	system("pause");
}

File: k-means.py
import numpy as np
import math

import matplotlib.pyplot as plt
import matplotlib.cm as cm

import csv

"""
#计算两个经纬度之间的距离
def distCalculation(vecA,vecB):
    Lat1=math.radians(vecA[0][0,1])
    Lat2=math.radians(vecB[0][0,1])
    a=Lat1-Lat2
    b=math.radians(vecA[0][0,0])-math.radians(vecB[0][0,0])
    s = 2 * math.asin(math.sqrt(math.pow(math.sin(a / 2), 2) + math.cos(Lat1)*math.cos(Lat2)*math.pow(math.sin(b / 2), 2)))
    distance=s*6378137
    return distance



# 构建聚簇中心，取k个随机质心
def randCent(dataSet, k):
    n = np.shape(dataSet)[1]
    centroids = np.mat(np.zeros((k,n)))   # 每个质心有n个坐标值，总共要k个质心
    for j in range(n):
        minJ = min(dataSet[:,j])
        maxJ = max(dataSet[:,j])
        rangeJ = float(maxJ - minJ)
        centroids[:,j] = minJ + rangeJ * np.random.rand(k, 1)
    return centroids


# 加载数据
def loadDataSet(fileName):  # 解析文件，得到一个浮点数字类型的矩阵
    totalDataMat = []              # 文件的最后一个字段是类别标签
    fr = open(fileName)
    for line in fr.readlines():
        curLine = line.strip('\n').split(',')
        fltLine = list(map(float, curLine))    # 将每个元素转成float类型
        totalDataMat.append(fltLine)
    return np.mat(totalDataMat)


# k-means 聚类算法
def kMeans(dataSet, k, distMeans =distCalculation, createCent = randCent):
    m = np.shape(dataSet)[0]
    clusterAssment = np.mat(np.zeros((m,2)))    # 用于存放该样本属于哪类及质心距离
    # clusterAssment第一列存放该数据所属的中心点，第二列是该数据到中心点的距离
    centroids = createCent(dataSet, k)
    clusterChanged = True   # 用来判断聚类是否已经收敛
    Terminator=[]
    for i in range(k):
        Terminator.append([])
    while clusterChanged:
        clusterChanged = False
        for i in range(m):  # 把每一个数据点划分到离它最近的中心点
            minDist = np.inf; minIndex = -1
            for j in range(k):
                distJI = distMeans(centroids[j,:], dataSet[i,:])
                if distJI < minDist:
                    minDist = distJI; minIndex = j  # 如果第i个数据点到第j个中心点更近，则将i归属为j
            Terminator[minIndex].append(i)

    return centroids,Terminator




# --------------------测试----------------------------------------------------
# 用测试数据及测试kmeans算法
datMat = loadDataSet('C:\\Users\\Rooobins\\Desktop\\operrecord_18062410.txt')
myCentroids,clustAssing = kMeans(datMat,500)



colors=iter(cm.rainbow(np.linspace(0,1,len(clustAssing))))
x=[]
y=[]
for listData in clustAssing:
    x.clear()
    y.clear()
    for i in listData:
        x.append(datMat[i,0])
        y.append(datMat[i,1])

    plt.scatter(x,y,color=next(colors))


plt.show()


writeClustAssing=open("C:\\Users\\Rooobins\\Desktop\\clustAssing.txt",'w+')
writeClustAssing.writelines(str(clustAssing))
writeClustAssing.close()

"""





with open("C:\\Users\\Rooobins\\Desktop\\data.csv",'w+')  as csvfile:
    import csv

    with open("C:\\Users\\Rooobins\\Desktop\\data.csv",'w+') as csvfile:
        spamwriter = csv.writer(csvfile, dialect='excel')
        # 读要转换的txt文件
        with open("C:\\Users\\Rooobins\\Desktop\\operrecord_18062410.txt", 'r+') as filein:
            for line in filein:
                line_list = line.strip().split()
                spamwriter.writerow(line_list)





